### ISSUE TITLE
saveAll() performance issues

### ISSUE THREAD

**Issue by pillvin** [ClaudioConsolmagno/spring-data-jpa-temporal#1]
SimpleJPARepository.saveAll() loops through iterator instance and invokes TemporalRepository.save() method.

save method is currently expensive because it makes 3 calls :-
1) findById -> db call
2) deleteById -> db call
3) insert-> db call

Invoking saveAll on 3000 objects takes almost 10 seconds per object OR 30,000 seconds = ~8.33 hours. The 10 seconds per object could be due to the fact that the table already has 5 million records.

It's be great if you could look at this aspect and provide a fix, or recommend a workaround to solve this performance issue.

thanks,
Vinod

**Comment by ClaudioConsolmagno** [comment]
Thanks for the feedback Vinod! You're right, `saveAll()` isn't very performant. I never looked in depth at performance on TemporalRepository methods but for this one I think there's a quick solution here to make great gains. I'll come back with something within the next day or two.

**Comment by ClaudioConsolmagno** [comment]
Hi Vinod, I have released version 0.1.5 with an overriden `saveAll()` method. Now, it should make at most 1 db call with findAllById, 1 db call with deleteByIds(), and, 1 save call for each entity being saved that has been changed. The trade of here is that we are loading up into memory the existing entities from the DB all in one go to compare with the ones being saved. IMO this is fine for the general use case.

From my test cases, I saw the same saveAll() on 30k entities taking 45 minutes to run before this update, and now taking less than 4 minutes.

I'm closing this unless there's any other thoughts on this matter, in which case I could re-open.
